### PR TITLE
8317368: [JVMCI] SIGSEGV in JVMCIEnv::initialize_installed_code on libgraal

### PR DESCRIPTION
**Pull request by tkrodriguez**
This fixes some lurking issues with JVMCI and nmethod related both BarrierSetNMethod and the garbage collection of nmethods.  In particular the stack walking in c2v_iterateFrames visits many frames and needs a KeepStackGCProcessedMark for safety.  Additionally, JVMCI interacts with nmethods in complex ways and needs some sort of strong root during these interactions.  A new JavaThread field has been added that mirrors the way JVMTI keeps nmethods alive.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317368](https://bugs.openjdk.org/browse/JDK-8317368): [JVMCI] SIGSEGV in JVMCIEnv::initialize_installed_code on libgraal (**Bug** - P3)

### Reviewers
 * [Doug Simon](https://openjdk.org/census#dnsimon) (@dougxc - **Reviewer**) ⚠️ Review applies to [429a769f](https://git.openjdk.org/jdk/pull/17714/files/429a769f5c4a0e78690c3c4a988a17d361a870b4)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [3d2417e7](https://git.openjdk.org/jdk/pull/17714/files/3d2417e75d2fba308e85e85b4807e282d172d25f)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [3d2417e7](https://git.openjdk.org/jdk/pull/17714/files/3d2417e75d2fba308e85e85b4807e282d172d25f)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17714/head:pull/17714` \
`$ git checkout pull/17714`

Update a local copy of the PR: \
`$ git checkout pull/17714` \
`$ git pull https://git.openjdk.org/jdk.git pull/17714/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17714`

View PR using the GUI difftool: \
`$ git pr show -t 17714`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17714.diff">https://git.openjdk.org/jdk/pull/17714.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17714#issuecomment-1928523128)

### CODE REVIEW

**Comment by dougxc** [comment]
I don't see an easy way to avoid the extra field in `JavaThread` given that any Java thread can install JVMCI code, not just CompilerThreads.

**Comment by dougxc** [comment]
Can this be merged soon Tom?

**Comment by tkrodriguez** [comment]
I resync'ed with master and mach5 testing was clean. I'll ping some likely reviewers.

**Review by vnkozlov — APPROVED** [review]
Seems fine.

May be we should consider specialized `jvmciJavaThread` subclass to keep all JVMCI fields instead of having them in `JavaThread` even when Graal is not used.

**Comment by vnkozlov** [comment]
@tkrodriguez  did you consider an nmethod's state similar to `not_installed` to avoid this issue?

**Review by fisk — APPROVED** [review]
Looks good.

File: src/hotspot/share/jvmci/jvmciCodeInstaller.cpp [hunk 1]
```diff
@@ -813,14 +816,12 @@ JVMCI::CodeInstallResult CodeInstaller::install(JVMCICompiler* compiler,
...
-        if (_nmethod_entry_patch_offset != -1) {
-          err_msg msg("");
-          BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
+      if (_nmethod_entry_patch_offset != -1) {
+        BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
 
-          if (!bs_nm->verify_barrier(nm, msg)) {
-            JVMCI_THROW_MSG_(IllegalArgumentException, err_msg("nmethod entry barrier is malformed: %s", msg.buffer()), JVMCI::ok);
-          }
+        err_msg msg("");
```
Comment (dougxc) [inline src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:822]: delete `err_msg msg("");`

**Comment by tkrodriguez** [comment]
You mean to add a level of indirection for the JVMCI specific fields?  Some of them would be fine with that but a few really want direct access.  We'd have to rework some interpreter assembly and code generation on the Graal side for those fields that accessed directly.  I could look into it at some point.

File: src/hotspot/share/jvmci/jvmciCodeInstaller.cpp (hunk 1 above)
Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:822]: It's necessary to allocate an empty error buffer for use by the verify_barrier code.  Is there some better way to express this?

File: src/hotspot/share/jvmci/jvmciEnv.hpp [hunk 2]
```diff
@@ -374,7 +389,7 @@ class JVMCIEnv : public ResourceObj {
   CodeBlob* get_code_blob(JVMCIObject code);
 
   // Given an instance of HotSpotInstalledCode return the corresponding nmethod.
```
[the remark this answers is not in the document]
Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmciEnv.hpp:391]: ok

File: src/hotspot/share/runtime/javaThread.hpp [hunk 3]
```diff
@@ -381,6 +381,9 @@ class JavaThread: public Thread {
   jlong      _jvmci_reserved1;
   oop        _jvmci_reserved_oop0;
 
+  // This field is used to keep an nmethod visible to the GC so that it can be kept alive
```
[the remark this answers is not in the document]
Comment (tkrodriguez) [reply src/hotspot/share/runtime/javaThread.hpp]: ok

File: src/hotspot/share/jvmci/jvmciCodeInstaller.cpp (hunk 1 above)
Comment (dougxc) [reply src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:822]: Oh, I missed that `msg` is used below. I don't know of a better way to do this. Maybe just add a comment:
```
// an empty error buffer for use by the verify_barrier code
```

Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:822]: ok

**Comment by tkrodriguez** [comment]
Regarding the comment about the special state, the problem is that the nmethod must be visited by the GC while it's alive and the only way to ensure that for it to be explicitly visited during GC.  There's no way to fixup the nmethod after the fact.

**Comment by tkrodriguez** [comment]
Thanks for the reviews.

/integrate

**Pull request closed without merging**